### PR TITLE
Enhancing Mocca API to define GraphQL variables #13 #14

### PR DESCRIPTION
**Pull request by fabiocarvalho777**

### CODE REVIEW

File: docs/END_USER_DOCUMENT.md [hunk 1]
```diff
@@ -193,20 +196,21 @@ You can see below an example of a simple client API.
...
-    List<Book> getBooks(String variables);
+    List<Book> getBooks(@Var("authorId") long authorId);
 
     @Query
-    Book getBook(long id);
+    Book getBook(@Var("id") long id);
 
     @Mutation
-    Author addAuthor(Author author);
+    Author addAuthor(@Var(value = "author", ignore = "books") Author author);
```
Comment (abhijith-prabhakar) [inline docs/END_USER_DOCUMENT.md:210]: what happens if 'books' property does not exists on Author?

File: mocca-client/src/main/java/com/paypal/mocca/client/MoccaFeignEncoder.java [hunk 2]
```diff
@@ -114,33 +121,38 @@ static SelectionSet getSelectionSet(RequestTemplate requestTemplate) {
...
+        List<Variable> variables = new ArrayList<>(parameters.length);
         for (int i = 0; i < parameters.length; i++) {
-            if (parameters[i].getAnnotation(RequestHeaderParam.class) == null) {
-                varCount++;
-                varIndex = i;
+            Parameter parameterMetadata = parametersMetadata[i];
+            Var varAnnotation = parameterMetadata.getAnnotation(Var.class);
+            // FIXME It would be better if this check happened at client definition time, instead of request time
+            if (varAnnotation == null) {
+                if (parameterMetadata.getAnnotation(RequestHeaderParam.class) == null) {
```
Comment (jamesmartinpp) [inline mocca-client/src/main/java/com/paypal/mocca/client/MoccaFeignEncoder.java:142]: This "if" block appears to be copied and could be extract outside the if/else block I think.

File: mocca-client/src/main/java/com/paypal/mocca/client/MoccaFeignEncoder.java [hunk 3]
```diff
@@ -114,33 +121,38 @@ static SelectionSet getSelectionSet(RequestTemplate requestTemplate) {
...
+                    final String method = requestTemplate.methodMetadata().method().getName();
+                    throw new MoccaException("Invalid GraphQL operation method " + method + ", make sure all its parameters are annotated with one Mocca annotation");
+                }
+            } else {
+                if (parameterMetadata.getAnnotation(RequestHeaderParam.class) != null) {
```
Comment (jamesmartinpp) [inline mocca-client/src/main/java/com/paypal/mocca/client/MoccaFeignEncoder.java:147]: same as above.

File: mocca-client/src/main/java/com/paypal/mocca/client/annotation/Var.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+     *
+     *     &#064;Query
+     *     &#064;SelectionSet("{id, name}")
+     *     List&#60;Book&#62; getBooks(&#064;Var("authorId") String query);
+     *
+     * </code></pre>
+     *
+     * @return whether all GraphQL variables should be set using one String parameter
+     */
+    // TODO This feature is not implemented yet
```
Comment (jamesmartinpp) [inline mocca-client/src/main/java/com/paypal/mocca/client/annotation/Var.java]: If the feature is not implemented perhaps this section should be removed.

File: mocca-client/src/main/java/com/paypal/mocca/client/annotation/Var.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    /**
+     * When the variable is a POJO (following Java bean conventions), sometimes it is populated with certain properties
+     * the application would like to be ignored by Mocca when serializing the GraphQL variables.
+     * In order to do so, specify here an array containing all fields to be ignored in the POJO object.
+     * The name of a property in an inner POJO can be specified using the outer field name followed by dot.
+     * If the request type is not a POJO, or if {@link #raw()} is set to true, then {@code ignore} will have no effect.
+     *
+     * @return an array containing all fields to be ignored in a POJO variable
+     */
+    String[] ignore() default "";
```
Comment (jamesmartinpp) [inline mocca-client/src/main/java/com/paypal/mocca/client/annotation/Var.java]: Should the default be null or []?

**Review by jamesmartinpp — COMMENTED** [review]
Just a few small comments and questions.

**Review by abhijith-prabhakar — APPROVED** [review]

File: mocca-client/src/main/java/com/paypal/mocca/client/MoccaFeignEncoder.java (hunk 2 above)
Comment (fabiocarvalho777) [reply mocca-client/src/main/java/com/paypal/mocca/client/MoccaFeignEncoder.java:142]: It is not, the condition is different.

File: mocca-client/src/main/java/com/paypal/mocca/client/MoccaFeignEncoder.java (hunk 3 above)
Comment (fabiocarvalho777) [reply mocca-client/src/main/java/com/paypal/mocca/client/MoccaFeignEncoder.java:147]: Same as last comment

File: mocca-client/src/main/java/com/paypal/mocca/client/annotation/Var.java (hunk 5 above)
Comment (fabiocarvalho777) [reply mocca-client/src/main/java/com/paypal/mocca/client/annotation/Var.java]: I tried that, but the compiler didn't like it. It has to be a constant.

File: mocca-client/src/main/java/com/paypal/mocca/client/annotation/Var.java (hunk 4 above)
Comment (fabiocarvalho777) [reply mocca-client/src/main/java/com/paypal/mocca/client/annotation/Var.java]: Thanks for catching this. This feature is already implemented, I just forgot to remove the TODO comment. I will remove it.

File: docs/END_USER_DOCUMENT.md (hunk 1 above)
Comment (fabiocarvalho777) [reply docs/END_USER_DOCUMENT.md:210]: Good point Abhi. Nothing happens. But still, that should be documented. I will document that in the annotation javadoc. Thanks.

File: mocca-client/src/main/java/com/paypal/mocca/client/annotation/Var.java (hunk 4 above)
Comment (fabiocarvalho777) [reply mocca-client/src/main/java/com/paypal/mocca/client/annotation/Var.java]: Done

File: docs/END_USER_DOCUMENT.md (hunk 1 above)
Comment (fabiocarvalho777) [reply docs/END_USER_DOCUMENT.md:210]: Done

File: mocca-client/src/main/java/com/paypal/mocca/client/annotation/Var.java (hunk 5 above)
Comment (fabiocarvalho777) [reply mocca-client/src/main/java/com/paypal/mocca/client/annotation/Var.java]: I have just tried `{}` and it worked. I updated the PR. Thanks.

File: mocca-client/src/main/java/com/paypal/mocca/client/MoccaFeignContract.java [hunk 6]
```diff
@@ -1,13 +1,7 @@
...
 
-import com.paypal.mocca.client.annotation.Mutation;
-import com.paypal.mocca.client.annotation.Query;
-import com.paypal.mocca.client.annotation.RequestHeader;
-import com.paypal.mocca.client.annotation.RequestHeaderParam;
-import feign.DeclarativeContract;
-import feign.MethodMetadata;
-import feign.Request;
-import feign.Util;
+import com.paypal.mocca.client.annotation.*;
```
Comment (jamesmartinpp) [inline mocca-client/src/main/java/com/paypal/mocca/client/MoccaFeignContract.java:3]: do we want to encourage wildcard imports? what's the policy?

**Review by jamesmartinpp — APPROVED** [review]
LGTM

**Pull request merged**